servicemanager/servicemanager.go

```diff
@@ -36,34 +46,211 @@ const maxConcurrentActions = 10
 	}
 
+	services, err := sm.serviceInfoProvider.GetAllServices()
+	if err != nil {
+		return nil, aoserrors.Wrap(err)
+	}
+
+	for _, service := range services {
+		if err = sm.idsPool.AddID(service.GID); err != nil {
+			log.Errorf("Can't add service UID/GID to pool: %s", err)
+		}
+	}
+
 	return sm, nil
 }
 
+// InstallService install service to the system.
+func (sm *ServiceManager) InstallService(
+	newService ServiceInfo, imageURL string, fileInfo image.FileInfo) (err error) {
+	return <-sm.actionHandler.Execute(newService.ServiceID,
+		func(id string) error {
+			return sm.doInstallService(newService, imageURL, fileInfo)
+		},
+	)
+}
+
 /**********************************************************************************************************************
  * Private
  *********************************************************************************************************************/
+
+func (sm *ServiceManager) doInstallService(newService ServiceInfo, imageURL string, fileInfo image.FileInfo) error {
+	serviceFromDB, err := sm.serviceInfoProvider.GetService(newService.ServiceID)
+	if err != nil && !errors.Is(err, ErrNotExist) {
+		return aoserrors.Wrap(err)
+	}
+
+	var isServiceExist bool
+
+	if err == nil {
+		isServiceExist = true
+
+		if newService.AosVersion <= serviceFromDB.AosVersion {
+			return ErrVersionMismatch
+		}
+	}
+
+	// download and unpack
+	urlVal, err := url.Parse(imageURL)
+	if err != nil {
+		return aoserrors.Wrap(err)
+	}
+
+	var serviceImage string
+
+	if urlVal.Scheme != "file" {
+		if serviceImage, err = image.Download(context.Background(), sm.downloadDir, imageURL); err != nil {
+			return aoserrors.Wrap(err)
+		}
+
+		defer os.RemoveAll(serviceImage)
+	} else {
+		serviceImage = urlVal.Path
+	}
+
+	if err = image.CheckFileInfo(context.Background(), serviceImage, fileInfo); err != nil {
+		return aoserrors.Wrap(err)
+	}
+
+	// create install dir
+	newService.ImagePath, err = ioutil.TempDir(sm.servicesDir, "")
+	if err != nil {
+		return aoserrors.Wrap(err)
+	}
+
+	if err = imageutils.UnpackTarImage(serviceImage, newService.ImagePath); err != nil {
+		return aoserrors.Wrap(err)
+	}
+
+	if err = validateUnpackedImage(newService.ImagePath); err != nil {
+		return aoserrors.Wrap(err)
+	}
+
+	// We need to install or update the service
+
+	defer func() {
+		if err != nil {
+			// Remove install dir if exists
+			if _, err := os.Stat(newService.ImagePath); err == nil {
+				_ = os.RemoveAll(newService.ImagePath)
+			}
+		}
+	}()
+
+	log.WithFields(log.Fields{"dir": newService.ImagePath, "serviceID": newService.ServiceID}).Debug("Create install dir")
+
+	if err = sm.prepareService(&newService, &serviceFromDB, isServiceExist); err != nil {
+		return aoserrors.Wrap(err)
+	}
+
+	if err = sm.serviceInfoProvider.AddService(newService); err != nil {
+		return aoserrors.Wrap(err)
+	}
+
+	return nil
+}
+
+func (sm *ServiceManager) prepareService(newService, oldService *ServiceInfo, update bool) (err error) {
```

Comment:
Could you set manifest and GID in the above function? Then this function can be renamed to `prepareServiceFS` for example. The main disadvantage in this implementation it is modifying the ServiceInfo var through the pointer.

Reply:
reworked